### ISSUE TITLE
[Proposal] Safer, cleaner, and simpler teardowns through generator fixtures

### ISSUE THREAD

**Issue by SalmonMode** [mochajs/mocha#4368]
Tl;dr: Added ability to use yield in before and beforeEach to separate setup and teardown logic. Teardowns will then only get executed if the setup doesn't have any issues.

## Example scenario

An e2e test needs to instantiate a client to interact with the admin API, use that client to create a user, launch a web browser session, use that session to go to a login page, and then login using the created user. In order to properly tear down after this test, the browser session must be quit, and the user must be deleted.

### The challenge introduced

While some parts of this setup process are independent of each other, there must be some serializable order to them; certain fixtures will have to execute/resolve before others. There is a chance that an error can occur in an earlier fixture that would prevent further fixtures from setting up.

Mocha currently handles this part just fine, as it will just stop executing `before` and `beforeEach` functions for that test.

The challenge is introduced with how teardowns can get handled. When mocha reaches a given `describe` layer for a test without any errors popping up before then, if a `before`/`beforeEach` throws an error, mocha will still try to execute all the `after`/`afterEach` for that layer, no matter how many `before`/`beforeEach` succeeded in that later. This presents a risk in that mocha could still attempt to tear down a fixture even if its setup wasn't completed.

Here's a quick mockup demonstrating the risks of this if the engineer assumes the setups won't have issues:

```javascript
describe('loggin in', () => {
  var driver;
  var loginPage;
  var landingPage;
  var adminClient;
  var user;

  before('create user', () => {
    user = User({ name = 'Jimmy', email = `test+${uuidv4()}@example.com`, password = 'P4$$w0rd!' });
    adminClient = AdminClient();
    adminClient.createUser(user);
  });

  before('launch driver', () => {
    driver = Chrome();
  });

  before('go to login page', () => {
    driver.get("https://mysite.com/login");
    loginPage = LoginPage(driver);
  });

  before('login', () => {
    loginPage.login(user);
    landingPage = LandingPage(driver);
  });

  it('should show user name in header', () => {
    expect(landingPage.header.text, 'to equal', user.name);
  });

  after('quit driver', () => {
    driver.quit();
  });

  after('delete user', () => {
    adminClient.deleteUser(user);
  });
});
```

If creating the user throws an error, the driver will never be launched, but mocha would still attempt to quit it in the first `after`. What's worse, is that if the driver `before` was the one to have the error, the `driver.quit()` call would also throw an error, as there's no driver to actually quit, and, as a result, mocha would not attempt to finish the remaining teardowns, which means the user wouldn't get deleted.

There is no _guaranteed_ solution for the latter risk, as even if the driver had launched successfully, it's still possible (however unlikely) that quitting that driver session could throw an error.

The prior risk, though, is at least manageable. By doing a preliminary check in the `after`/`afterEach` to see if there is something that needs to be torn down would mitigate this risk. The `before` for user creation would have to be modified to this:

```javascript
  before('create user', () => {
    let _user = User({ name = 'Jimmy', email = `test+${uuidv4()}@example.com`, password = 'P4$$w0rd!' });
    adminClient = AdminClient();
    adminClient.createUser(_user);
    // necessary to delay assignment until user creation completion for simpler teardown logic
    user = _user;
  });
```

and the `after` for the driver and user creation would look like this:

```javascript
  after('quit driver', () => {
    if (driver) {
      driver.quit();
    }
  });

  after('delete user', () => {
    if (user) {
      adminClient.deleteUser(user);
    }
  });
```

This, however, increases the [complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of the tests, which would be ideal to avoid. Depending on how complex it is to either check for a need to teardown a fixture, or to do the teardown itself, this can drastically increase the complexity and the surface area for bugs in the test code (although, ideally for maximum safety/simplicity, each fixture would be atomic so the resulting complexity would be minimal).

## Proposed solution

Rather than having increased complexity in the test code to handle conditionally tearing down, it would be beneficial to not have the conditional at all, thus reducing complexity. Going one step further, it would also be beneficial to explicitly tie the teardown logic for a fixture to its setup logic in such a way that the teardown logic _can't_ be executed if the associated setup logic had a problem, thus reducing complexity further and increasing teardown safety.

By making fixtures into generators this can be easily achieved. I'm told generators aren't widely used for JavaScript development, but in this use case, they become more like context managers than generators.

Here's what the above example would look like using them:

```javascript
describe('loggin in', () => {
  var driver;
  var loginPage;
  var landingPage;
  var adminClient;
  var user;

  before('create user', function*() {
    user = User({ name = 'Jimmy', email = `test+${uuidv4()}@example.com`, password = 'P4$$w0rd!' });
    adminClient = AdminClient();
    adminClient.createUser(u);
    yield;
    adminClient.deleteUser(user);
  });

  before('launch driver', function*() {
    driver = Chrome();
    yield;
    driver.quit();
  });

  before('go to login page', () => {
    driver.get("https://mysite.com/login");
    loginPage = LoginPage(driver);
  });

  before('login', () => {
    loginPage.login(user);
    landingPage = LandingPage(driver);
  });

  it('should show user name in header', () => {
    expect(landingPage.header.text, 'to equal', user.name);
  });
});
```

In my implementation, mocha evaluates a hook's result after executing it initially and calling `next()` on it once. If the hook made it through to the `yield` successfully, then it is appended to the relevant `after`/`afterEach` array for that `describe` layer. Because they are executed sequentially (assuming the fixtures are executed in a serializable fashion), then the teardowns for them will be executed in the reverse order, effectively creating a fixture stack. So because the user was created _before_ the driver was launched, the user would be deleted _after_ the driver was quit.

If a hook throws an error during its setup logic, it will not be added the the relevant `after`/`afterEach` array, and as a result, mocha won't try to execute its teardown code, but any hooks that had already been setup successfully will still be torn down.

There's also the bonus of not having to worry about the order of the `after`/`afterEach` definitions. for a given layer, which increases test code maintainability/extensibility/reusability.

[WIP PR is here.](https://github.com/mochajs/mocha/pull/4369)
